README.md
Normalized Lua Functions
========================

Copyright (C) 2002-2022 [std.normalize authors][authors]

[![License](https://img.shields.io/:license-mit-blue.svg)](https://mit-license.org)
[![workflow status](https://github.com/lua-stdlib/normalize/actions/workflows/spec.yml/badge.svg?branch=master)](https://github.com/lua-stdlib/normalize/actions)
[![codecov.io](https://codecov.io/gh/lua-stdlib/normalize/branch/master/graph/badge.svg)](https://codecov.io/gh/lua-stdlib/normalize)

This is a collection of normalized lua functions for Lua 5.1 (including
LuaJIT), 5.2, 5.3 and 5.4.  The libraries are copyright by their authors
(see the [AUTHORS][] file for details), and released under the
[MIT license][mit] (the same license as Lua itself).  There is no
warranty.

_normalize_ has no run-time prerequisites beyond a standard Lua system,
and the [std._debug][] package - though it will also take advantage of
[std.strict][] if installed.

Writing Lua libraries that target several Lua implementations can be a
frustrating exercise in working around lots of small differences in APIs
and semantics they share (or rename, or omit).  _normalize_ provides the
means to simply access deterministic implementations of those APIs while
preserving the the same semantics across all supported host Lua
implementations.  Each function is as thin and fast an implementation as
possible within that host Lua environment, evaluating to the Lua C
implementation with no overhead where host semantics allow.

The core of this module is to transparently set the environment up with
a single API (as opposed to requiring caching functions from a module
table into module locals):

```lua
   local _ENV = require 'std.normalize' {
     'package',
     'std.prototype',
     strict = 'std.strict',
   }
   
   -- From here pairs, ipairs, setfenv et. al. work the same everywhere!
   -- While 'package' is imported, 'debug', 'table', et. al. are not
   -- preventing undeclared use.  External modules can be loaded into
   -- nested tables (std.prototype) or specific symbols (strict).
```

The normalized API is not yet complete, and in contrast to the
[lua-compat][] libraries, neither does it attempt to provide you with as 
nearly compatible an API as is possible relative to some specific Lua
implementation - rather it provides a variation of the "lowest common
denominator" that can be implemented relatively efficiently in the
supported Lua implementations. At the moment, only the functionality
required by [stdlib][] is provided.

Pull-requests for more normalized APIs are welcome!

[authors]: https://github.com/lua-stdlib/normalize/blob/master/AUTHORS.md
[github]: https://github.com/lua-stdlib/normalize/ "Github repository"
[lua]: https://www.lua.org "The Lua Project"
[lua-compat]: https://github.com/keplerproject/lua-compat-5.3 "Lua 5.3ish API"
[mit]: https://mit-license.org "MIT License"
[stdlib]: https://github.com/lua-stdlib/lua-stdlib "Standard Lua Libraries"
[std.strict]: https://github.com/lua-stdlib/strict "strict variables"
[typecheck]: https://github.com/gvvaughan/typecheck "function type checks"


Installation
------------

The simplest and best way to install normalize is with [LuaRocks][]. To
install the latest release (recommended):

```bash
    luarocks install std.normalize
```

To install current git master (for testing, before submitting a bug
report for example):

```bash
    luarocks install https://raw.githubusercontent.com/lua-stdlib/normalize/master/normalize-git-1.rockspec
```

The best way to install without [LuaRocks][] is to copy the `std/normalize`
folder and its contents into a directory on your package search path.

[luarocks]: https://www.luarocks.org "Lua package manager"


Documentation
-------------

The latest release of these libraries is [documented in LDoc][github.io].
Pre-built HTML files are included in the release.

[github.io]: https://lua-stdlib.github.io/normalize


Bug reports and code contributions
----------------------------------

These libraries are written and maintained by their users.

Please make bug reports and suggestions as [GitHub Issues][issues].
Pull requests are especially appreciated.

But first, please check that your issue has not already been reported by
someone else, and that it is not already fixed by [master][github] in
preparation for the next release (see Installation section above for how
to temporarily install master with [LuaRocks][]).

There is no strict coding style, but please bear in mind the following
points when proposing changes:

0. Follow existing code. There are a lot of useful patterns and avoided
   traps there.

1. 3-character indentation using SPACES in Lua sources: It makes rogue
   TABS easier to see, and lines up nicely with 'if' and 'end' keywords.

2. Simple strings are easiest to type using single-quote delimiters,
   saving double-quotes for where a string contains apostrophes.

3. Save horizontal space by only using SPACES where the parser requires
   them.

4. Use vertical space to separate out compound statements to help the
   coverage reports discover untested lines.

[issues]: https://github.com/lua-stdlib/normalize/issues


NEWS.md
# std.normalize NEWS - User visible changes

## Noteworthy changes in release ?.? (????-??-??) [?]

### New features

  - Support importing into another project directly with:

    ```sh
    $ cp ../normalize/lib/std/normalize/init.lua lib/std/normalize.lua
    ```


## Noteworthy changes in release 2.0.3 (2020-04-15) [stable]

### New features

  - Initial support for Lua-5.4.


## Noteworthy changes in release 2.0.2 (2018-03-17) [stable]

### Bug fixes

  - Passing a table to the return value of `require 'std.normalize'`
    really does diagnose attempts to access undeclared variables when
    std.strict is found in the package.path now.

  - Don't trigger undeclared variable accesses during loading, when
    used in conjunction with other strict modules.


## Noteworthy changes in release 2.0.1 (2017-11-26) [stable]

### Bug fixes

  - Environment table population now correctly looks up ALLCAPS symbols
    when inferring destination symbol names, and only treats explicit
    assignment to an ALLCAPS destination symbol name as a string
    constant:

    ```lua
    local _ENV = require 'std.normalize' {
       'lyaml.functional.NULL',
       CONST = 'lyaml.functional.NULL',
    }
    assert(NULL == require 'lyaml.functional'.NULL)
    assert(type(CONST) == 'string')
    ```

    Previously, `NULL` would not have been loaded, instead remaining as
    the constant string 'lyaml.functional.NULL'.


## Noteworthy changes in release 2.0 (2017-10-16) [stable]

### New features

  - New `string.render` exposes the low-level implementation of `str`.

  - New `table.keys` fetches a list of keys in a given table.

  - New `table.merge` performs destructive merging of table content.

### Bug fixes

  - Include `_G` table, pointing to the complete normalized table
    of loaded modules, in the default user environment.  So this
    will work now:

    ```lua
    local _ENV = require 'std.normalize' {}
    for i = 1, len(_G.arg) do
       print(i .. ': ' .. _G.arg[i])
    end
    ```

### Incompatible changes

  - Strict mode and argument checking are no longer controlled by the
    global _DEBUG setting, but rather by the new `std._debug` module.
    For example, to turn off all runtime debug facilities, insteod of
    setting `_DEBUG = false` before loading this module, you should use:

    ```lua
    local _debug = require 'std._debug'(false)
    ```

  - Removed ugly `opairs` implementation, with it's own fixed ordering.
    If you still need it, you can use something like this:

    ```lua
    keylist = table.keys(t)
    sort(keylist, function(a, b)
       if type(a) == 'number' then
          return type(b) ~= 'number' or a < b
       else
          return type(b) ~= 'number' and tostring(a) < tostring(b)
       end
    end)
    for _, k in ipairs(keylist) do
       process(k, t[k])
    end
    ```


## Noteworthy changes in release 1.0.4 (2017-09-11) [stable]

### New features

  - `std.normalize` has been rewritten, in addition to maintaining
    backwards compatibility, also allowing easier loading of symbols
    into the module environment:

    ```lua
    local _ENV = require 'std.normalize' {
       -- local string = require 'std.normalize.string'
       'string',
       -- local math = require 'math'
       math = math,
       -- local ceil = require 'math'.ceil
       ceil = math.ceil,
       -- local floor = require 'std.normalize.math'.floor
       'math.floor',
       -- local int = require 'std.normalize.math'.tointeger
       int = 'math.tointeger',
       -- local MODNAME = 'math.tointeger'
       MODNAME = 'math.tointeger',
    }
    ```

    Note that dot-delimited strings are searched in the 'std.normalize'
    table, and can be followed by optional nested table references to
    drill into that table.  Otherwise the references are filled from
    the host Lua _G as the table is populated before it is even passed
    to the 'std.normalize' loader.


## Noteworthy changes in release 1.0.3 (2017-09-02) [stable]

### New features

  - `std.normalize` allows assignment of non-string values in its table
    argument to propagate up to the assigned environment:

    ```lua
    local _ENV = require 'std.normalize' {
       ceil = math.ceil,
       argcheck = require 'typecheck'.argcheck,
    }
    ```

### Incompatible changes

  - `math.tointeger` returns `nil` for string type arguments now, for
    deterministic behaviour whether or not LUA_NOCVTS2N is defined at
    compile time.

  - `getfenv`, consequently, now diagnoses a bad argument when passed an
    integer-like string instead of an actual integer.


## Noteworthy changes in release 1.0.2 (2017-07-07) [stable]

### New features

  - `pack` sets a `__len` metamethod on its result so that `len` returns
    the actual number of arguments packed (including `nil`s).

  - `unpack`ing a `pack`ed sequence returns the original sequence without
    requiring an explicit `to_index` argument, even if the original
    sequence contains `nil`s:

    ```lua
    a, b, c = unpack(pack(1, nil, 3))
    assert(a == 1 and b == nil and c == 3, "require 'std.normalize' first!")
    ```

  - `str` uses C-like escape sequences \a, \b, \t, \n, \v, \f, \r and \\
    to render the associated bytes.

### Bug fixes

  - `getmetamethod` no longer raises an argerror for nil-valued
    initial argument.

  - `ipairs` and `opairs` now diagnose all non-table valued arguments
     correctly.

  - `str` now only skips consecutive integer keys in proper sequences.

### Incompatible changes

  - `ipairs` now respects the `__len` metamethod, such as the one set by
    `pack`.


## Noteworthy changes in release 1.0.1 (2016-05-28) [stable]

### Bug fixes

  - `argerror` correctly treats missing or `nil`-valued argument
    as equivalent to level 1 (i.e. blame the caller of the function
    that calls `argerror`).


## Noteworthy changes in release 1.0 (2016-05-08) [stable]

### New features (since lua-stdlib-41.2)


spec/spec_helper.lua
--[[
 Normalized Lua API for Lua 5.1, 5.2, 5.3 & 5.4
 Copyright (C) 2014-2023 std.normalize authors
]]
local typecheck
have_typecheck, typecheck = pcall(require, 'typecheck')

have_std_strict = pcall(require, 'std.strict.init')

local inprocess = require 'specl.inprocess'
local hell = require 'specl.shell'

badargs = require 'specl.badargs'

local cwd = os.getenv 'PWD'

package.path = cwd .. '/lib/?.lua;' .. cwd .. '/lib/?/init.lua;' .. package.path


-- Allow user override of LUA binary used by hell.spawn, falling
-- back to environment PATH search for 'lua' if nothing else works.
local LUA = os.getenv 'LUA' or 'lua'


-- Allow use of bare 'pack' and 'unpack' even in Lua 5.3.
pack = (function(f)
   local pack_mt = {
      ___len = function(self)
         return self.n
      end,
   }

   local pack_fn = f or function(...)
      return {n=select('#', ...), ...}
   end

   return function(...)
      return setmetatable(pack_fn(...), pack_mt)
   end
end)(rawget(_G, "pack"))

unpack = table.unpack or unpack


function callable(x)
   if type(x) == 'function' or (getmetatable(x) or {}).__call then
      return x
   end
end


function getmetamethod(x, n)
   return callable((getmetatable(x) or {})[n])
end


function copy(t)
   local r = {}
   for k, v in next, t do r[k] = v end
   return r
end


-- In case we're not using a bleeding edge release of Specl...
_diagnose = badargs.diagnose
badargs.diagnose = function(...)
   if have_typecheck then
      return _diagnose(...)
   end
end


-- A copy of base.lua:type, so that an unloadable base.lua doesn't
-- prevent everything else from working.
function objtype(o)
   return(getmetatable(o) or {})._type or io.type(o) or type(o)
end


local function mkscript(code)
   local f = os.tmpname()
   local h = io.open(f, 'w')
   -- TODO: Move this into specl, or expose arguments so that we can
   --          turn this on and off based on specl `--coverage` arg.
   h:write "pcall(require, 'luacov')"
   h:write(code)
   h:close()
   return f
end


--- Run some Lua code with the given arguments and input.
-- @string code valid Lua code
-- @tparam[opt={}] string|table arg single argument, or table of
--    arguments for the script invocation.
-- @string[opt] stdin standard input contents for the script process
-- @treturn specl.shell.Process|nil status of resulting process if
--    execution was successful, otherwise nil
function luaproc(code, arg, stdin)
   local f = mkscript(code)
   if type(arg) ~= 'table' then arg = {arg} end
   local cmd = {LUA, f, unpack(arg)}
   -- inject env and stdin keys separately to avoid truncating `...` in
   -- cmd constructor
   cmd.env = { LUA_PATH=package.path, LUA_INIT='', LUA_INIT_5_2='' }
   cmd.stdin = stdin
   local proc = hell.spawn(cmd)
   os.remove(f)
   return proc
end


local function tabulate_output(code)
   local proc = luaproc(code)
   if proc.status ~= 0 then return error(proc.errout) end
   local r = {}
   proc.output:gsub('(%S*)[%s]*',
      function(x)
         if x ~= '' then r[x] = true end
      end)
   return r
end


--- Show changes to tables wrought by a require statement.
-- There are a few modes to this function, controlled by what named
-- arguments are given.   Lists new keys in T1 after `require "import"`:
--
--       show_apis {added_to=T1, by=import}
--
-- @tparam table argt one of the combinations above
-- @treturn table a list of keys according to criteria above
function show_apis(argt)
   return tabulate_output([[
      local before, after = {}, {}
      for k in pairs(]] .. argt.added_to .. [[) do
         before[k] = true
      end

      local M = require ']] .. argt.by .. [['
      for k in pairs(]] .. argt.added_to .. [[) do
         after[k] = true
      end

      for k in pairs(after) do
         if not before[k] then print(k) end
      end
   ]])
end


--[[ ================ ]]--
--[[ Tmpfile manager. ]]--
--[[ ================ ]]--


local tmpdir = string.gsub(
   os.getenv 'TMPDIR' or os.getenv 'TMP' or '/tmp',
   '([^/])/*$', '%1'
) .. '/normalize-' .. math.random(65536)


local function append(path, ...)
   local n = select('#', ...)
   if n > 0 then
      local fh = io.open(path, 'a')
      fh:write(table.concat({...}, '\n') .. '\n')
      fh:close()
   end
   return n
end


-- Create a temporary file.
-- @usage
--    h = Tmpfile()            -- empty generated filename
--    h = Tmpfile(content) -- save *content* to generated filename
--    h = Tmpfile(name, line, line, line) -- write *line*s to *name*
Tmpfile = setmetatable({}, {
   _type = 'Tmpfile',

   __call = function(self, path_or_content, ...)
      local new = {}
      if select('#', ...) == 0 then
         new.path = os.tmpname()
         append(new.path, path_or_content)
      else
         new.path = path_or_content
         append(new.path, ...)
      end
      return setmetatable(new, getmetatable(self))
   end,

   __index = {
      dirname = function(self)
         return self.path:gsub('/[^/]*$', '', 1)
      end,

      basename = function(self)
         return self.path:gsub('.*/', '')
      end,

      append = function(self, ...)
         return append(self.path, ...)
      end,

      remove = function(self)
         return os.remove(self.path)
      end,
   },
})


Tmpdir = setmetatable({}, {
   _type = 'Tmpdir',

   __call = function(self, dirname)
      local new = {
         path = dirname or tmpdir,
         children = {},
      }
      os.execute("mkdir '" .. new.path .. "'")
      return setmetatable(new, getmetatable(self))
   end,

   __index = {
      file = function(self, name, ...)
         local child = Tmpfile(self.path .. '/' .. name, ...)
         self.children[#self.children + 1] = child
         return child
      end,

      subdir = function(self, name)
         local child = Tmpdir(self.path .. '/' .. name)
         self.children[#self.children + 1] = child
         return child
      end,

      remove = function(self)
         for _, child in ipairs(self.children) do
	child:remove()
         end
         return os.remove(self.path)
      end,
   }
})



lib/std/normalize/_base.lua
--[[
 Normalized Lua API for Lua 5.1, 5.2, 5.3 & 5.4
 Coryright (C) 2014-2023 std.normalize authors
]]
--[[--
 Purely to break internal dependency cycles without introducing
 multiple copies of base functions used in other normalize modules.

 @module std.normalize._base
]]


local _ENV = require 'std.normalize._strict' (_G)

local getmetatable = getmetatable
local pack = table.pack or false
local tointeger = math.tointeger or false



--[[ =============== ]]--
--[[ Implementation. ]]--
--[[ =============== ]]--


local function getmetamethod(x, n)
   local m = (getmetatable(x) or {})[tostring(n)]
   if type(m) == 'function' then
      return m
   end
   if type((getmetatable(m) or {}).__call) == 'function' then
      return m
   end
end


local pack_mt = {
   __len = function(self)
      return self.n
   end,
}


local pack = pack or function(...)
   return {n=select('#', ...), ...}
end


local tointeger = (function(f)
   if not f then
      -- No host tointeger implementation, use our own.
      local floor = math.floor
      return function(x)
        if type(x) == 'number' and x - floor(x) == 0.0 then
           return x
        end
      end

   elseif f '1' ~= nil then
      -- Don't perform implicit string-to-number conversion!
      return function(x)
         if type(x) == 'number' then
            return tointeger(x)
         end
      end
   end

   -- Host tointeger is good!
   return f
end)(tointeger)



--[[ ================= ]]--
--[[ Public Interface. ]]--
--[[ ================= ]]--


return {
   --- Return named metamethod, if callable, otherwise `nil`.
   -- @see std.normalize.getmetamethod
   getmetamethod = getmetamethod,

   --- Return a list of given arguments, with field `n` set to the length.
   -- @see std.normalize.pack
   pack = function(...)
      return setmetatable(pack(...), pack_mt)
   end,

   --- Convert to an integer and return if possible, otherwise `nil`.
   -- @see std.normalize.math.tointeger
   tointeger = tointeger,
}


lib/std/normalize/_debug.lua
--[[
 Normalized Lua API for Lua 5.1, 5.2 & 5.3
 Coryright (C) 2014-2017 Gary V. Vaughan
]]
--[[--
 Depending on whether `std.debug_init` is installed, and on what the
 value of `_DEBUG` is in the global environment, return a table with
 a local `_DEBUG` that has at least `strict` and `argcheck` fields set
 appropriately.

 @module std.normalize._debug
]]

local _ENV = {
   _G = {
      _DEBUG = rawget(_G, '_DEBUG'),
   },
   next = next,
   pcall = pcall,
   require = require,
   setfenv = setfenv or function() end,
   type	 = type,
}
setfenv(1, _ENV)



--[[ =============== ]]--
--[[ Implementation. ]]--
--[[ =============== ]]--


local _DEBUG
do
   local ok, debug_init	= pcall(require , 'std.debug_init')
   if ok then
      -- Use the _DEBUG table from `std.debug_init`, if installed.
      _DEBUG = debug_init._DEBUG
   else
      local function choose(t)
         for k, v in next, t do
	if _G._DEBUG == false then
	   t[k] = v.fast
	elseif _G._DEBUG == nil then
	   t[k] = v.default
	elseif type(_G._DEBUG) ~= 'table' then
	   t[k] = v.safe
	elseif _G._DEBUG[k] ~= nil then
	   t[k] = _G._DEBUG[k]
	else
	   t[k] = v.default
	end
         end
         return t
      end

      _DEBUG = choose {
         argcheck = {default = true, safe = true, fast = false},
         strict	= {default = true, safe = true, fast = false},
      }
   end
end


--[[ ================= ]]--
--[[ Public Interface. ]]--
--[[ ================= ]]--


return _DEBUG


lib/std/normalize/_strict.lua
--[[
 Normalized Lua API for Lua 5.1, 5.2, 5.3 & 5.4
 Copyright (C) 2014-2022 std.normalize authors
]]
--[[--
 Depending on whether `std.strict` is installed, and the value of
 `std._debug.strict`, return a function for setting up a strict lexical
  environment for the caller.

 @module std.normalize._strict
]]

local setfenv = rawget(_G, 'setfenv') or function() end


local _ENV = {
   _debug = require 'std._debug',
   pcall = pcall,
   require = require,
   setmetatable = setmetatable,
}
setfenv(1, _ENV)



--[[ =============== ]]--
--[[ Implementation. ]]--
--[[ =============== ]]--


-- If strict mode is required, use 'std.strict' if we have it.
local strict
if _debug.strict then
   -- `require 'std.strict'` will get the old stdlib implementation of
   -- strict, which doesn't support environment tables :(
   ok, strict = pcall(require, 'std.strict.init')
   if not ok then
      strict = false
   end
end


--[[ ================= ]]--
--[[ Public Interface. ]]--
--[[ ================= ]]--


return setmetatable({
   --- Set a module environment, using std.strict if available.
   --
   -- Either 'std.strict' when available, otherwise a(Lua 5.1 compatible)
   -- function to set the specified module environment.
   -- @function strict
   -- @tparam table env module environment table
   -- @treturn table *env*, which must be assigned to `_ENV`
   -- @usage
   --    local _ENV = require 'std.normalize._strict'.strict {}
   strict = strict or function(env)
      return env
   end,
}, {
   --- Module Metamethods
   -- @section modulemetamethods

   --- Set a module environment, using std.strict if available.
   -- @function strict:__call
   -- @tparam table env module environment table
   -- @tparam[opt=1] int level stack level for `setfenv`, 1 means set
   --    caller's environment
   -- @treturn table *env*, which must be assigned to `_ENV`
   -- @usage
   --    local _ENV = require 'std.normalize._strict' {}
   __call = function(self, env, level)
      env = self.strict(env)
      setfenv(1+(level or 1), env)
      return env
   end,
})


specs/spec_helper.lua
--[[
 Normalized Lua API for Lua 5.1, 5.2 & 5.3
 Copyright (C) 2014-2017 Gary V. Vaughan
]]
local typecheck
have_typecheck, typecheck	= pcall (require, "typecheck")

local inprocess			= require "specl.inprocess"
local hell			= require "specl.shell"
local std			= require "specl.std"

badargs				= require "specl.badargs"

package.path = std.package.normalize ("./lib/?.lua", "./lib/?/init.lua", package.path)


-- Allow user override of LUA binary used by hell.spawn, falling
-- back to environment PATH search for "lua" if nothing else works.
local LUA = os.getenv "LUA" or "lua"


-- Allow use of bare 'pack' and 'unpack' even in Lua 5.3.
pack = table.pack or function (...) return {n = select ("#", ...), ...} end
unpack = table.unpack or unpack


local function getmetamethod (x, n)
  local m = (getmetatable (x) or {})[tostring (n)]
  if type (m) == "function" then
    return m
  end
  if type ((getmetatable (m) or {}).__call) == "function" then
    return m
  end
end


function callable (x)
  return type (x) == "function" or getmetamethod (x, "__call")
end


function copy (t)
  local r = {}
  for k, v in next, t do r[k] = v end
  return r
end


-- In case we're not using a bleeding edge release of Specl...
_diagnose = badargs.diagnose
badargs.diagnose = function (...)
  if have_typecheck then
    return _diagnose (...)
  end
end


-- A copy of base.lua:type, so that an unloadable base.lua doesn't
-- prevent everything else from working.
function objtype (o)
  return (getmetatable (o) or {})._type or io.type (o) or type (o)
end


local function mkscript (code)
  local f = os.tmpname ()
  local h = io.open (f, "w")
  -- TODO: Move this into specl, or expose arguments so that we can
  --       turn this on and off based on specl `--coverage` arg.
  h:write 'pcall (require, "luacov")'
  h:write (code)
  h:close ()
  return f
end


--- Run some Lua code with the given arguments and input.
-- @string code valid Lua code
-- @tparam[opt={}] string|table arg single argument, or table of
--   arguments for the script invocation.
-- @string[opt] stdin standard input contents for the script process
-- @treturn specl.shell.Process|nil status of resulting process if
--   execution was successful, otherwise nil
function luaproc (code, arg, stdin)
  local f = mkscript (code)
  if type (arg) ~= "table" then arg = {arg} end
  local cmd = {LUA, f, unpack (arg)}
  -- inject env and stdin keys separately to avoid truncating `...` in
  -- cmd constructor
  cmd.env = { LUA_PATH=package.path, LUA_INIT="", LUA_INIT_5_2="" }
  cmd.stdin = stdin
  local proc = hell.spawn (cmd)
  os.remove (f)
  return proc
end


local function tabulate_output (code)
  local proc = luaproc (code)
  if proc.status ~= 0 then return error (proc.errout) end
  local r = {}
  proc.output:gsub ("(%S*)[%s]*",
    function (x)
      if x ~= "" then r[x] = true end
    end)
  return r
end


--- Show changes to tables wrought by a require statement.
-- There are a few modes to this function, controlled by what named
-- arguments are given.  Lists new keys in T1 after `require "import"`:
--
--     show_apis {added_to=T1, by=import}
--
-- @tparam table argt one of the combinations above
-- @treturn table a list of keys according to criteria above
function show_apis (argt)
  return tabulate_output ([[
    local before, after = {}, {}
    for k in pairs (]] .. argt.added_to .. [[) do
      before[k] = true
    end

    local M = require "]] .. argt.by .. [["
    for k in pairs (]] .. argt.added_to .. [[) do
      after[k] = true
    end

    for k in pairs (after) do
      if not before[k] then print (k) end
    end
  ]])
end


--[[ ================ ]]--
--[[ Tmpfile manager. ]]--
--[[ ================ ]]--


local tmpdir = string.gsub (
  os.getenv "TMPDIR" or os.getenv "TMP" or "/tmp",
  "([^/])/*$", "%1"
) .. "/normalize-" .. math.random (65536)


local function append (path, ...)
  local n = select ("#", ...)
  if n > 0 then
    local fh = io.open (path, "a")
    fh:write (table.concat ({...}, "\n") .. "\n")
    fh:close ()
  end
  return n
end


-- Create a temporary file.
-- @usage
--   h = Tmpfile ()        -- empty generated filename
--   h = Tmpfile (content) -- save *content* to generated filename
--   h = Tmpfile (name, line, line, line) -- write *line*s to *name*
Tmpfile = setmetatable ({}, {
  _type = "Tmpfile",

  __call = function (self, path_or_content, ...)
    local new = {}
    if select ("#", ...) == 0 then
      new.path = os.tmpname ()
      append (new.path, path_or_content)
    else
      new.path = path_or_content
      append (new.path, ...)
    end
    return setmetatable (new, getmetatable (self))
  end,

  __index = {
    dirname = function (self)
      return self.path:gsub ("/[^/]*$", "", 1)
    end,

    basename = function (self)
      return self.path:gsub (".*/", "")
    end,

    append = function (self, ...)
      return append (self.path, ...)
    end,

    remove = function (self)
      return os.remove (self.path)
    end,
  },
})


Tmpdir = setmetatable ({}, {
  _type = "Tmpdir",

  __call = function (self, dirname)
    local new = {
      path = dirname or tmpdir,
      children = {},
    }
    os.execute ("mkdir '" .. new.path .. "'")
    return setmetatable (new, getmetatable (self))
  end,

  __index = {
    file = function (self, name, ...)
      local child = Tmpfile (self.path .. "/" .. name, ...)
      self.children[#self.children + 1] = child
      return child
    end,

    subdir = function (self, name)
      local child = Tmpdir (self.path .. "/" .. name)
      self.children[#self.children + 1] = child
      return child
    end,

    remove = function (self)
      for _, child in ipairs (self.children) do
	child:remove ()
      end
      return os.remove (self.path)
    end,
  }
})



--[[ ================ ]]--
--[[ Custom Matchers. ]]--
--[[ ================ ]]--


do
  -- Custom matcher for set size and set membership.

  local util     = require "specl.util"
  local matchers = require "specl.matchers"

  local Matcher, matchers, q =
        matchers.Matcher, matchers.matchers, matchers.stringify

  matchers.have_member = Matcher {
    function (self, actual, expect)
      return actual[expect] ~= nil
    end,

    actual = "set",

    format_expect = function (self, expect)
      return " a set containing " .. q (expect) .. ", "
    end,

    format_any_of = function (self, alternatives)
      return " a set containing any of " ..
             util.concat (alternatives, util.QUOTED) .. ", "
    end,
  }

  -- Alias that doesn't tickle sc_error_message_uppercase.
  matchers.raise = matchers.error
end
